src/mender-configure

```diff
@@ -2,13 +2,69 @@
 
 set -e
 
+CONFIG=${TEST_CONFIG:-/var/lib/mender-configure/device-config.json}
+APPLY_CMD=${TEST_APPLY_CMD:-/usr/lib/mender-configure/apply-device-config}
+
 STATE="$1"
 FILES="$2"
 
+BACKUP="$FILES/device-config-backup.json"
+
+safe_copy() {
+    if [ -n "$3" ]; then
+        echo "safe_copy can only handle one file copy at a time" 1>&2
+        exit 2
+    fi
+    cp "$1" "$2".tmp
+    sync "$2".tmp
+    mv "$2".tmp "$2"
+    sync "$(dirname "$2")"
```

Comment:
Why in the first case we sync the file while in the second case we sync the dir?

Reply:
The first call syncs the file content. This guarantees that the file contains the content, the whole content, and nothing but the content. The second call syncs the directory, IOW the rename, this guarantees that the file now occupies that file name, and isn't still using the tmp filename.

The reason they are done in stages and not together is that there is no guarantee about ordering when you sync both content and directory at the same time. It could turn out correct, but it could also decide to sync the directory first, and the content afterwards, which means a corrupt file will briefly occupy that name. If a spontaneous reboot happens at this moment, you are in trouble. The key guarantee we are taking advantage of here is that renames on the same filesystem are atomic, this is guaranteed by ext3 and ext4 (and most other journaled filesystems, but there are some exceptions).

Reply:
Thank you very much for the detailed explanation. Got it.